### PR TITLE
Issue #14600: Fix a performance degradation that was introduced when …

### LINKED ISSUE #14600: microprofile-primitives performance throughput regression
I am seeing a small regression in 20.0.0.12 with some of the microprofile-primitive runs.

A profile of the java code isn't showing me too much, but one area where it got slower is here:

```
  Parent      0   0.38   0.63          26          43    J:com/ibm/wsspi/webcontainer/collaborator/CollaboratorHelper.preInvokeCollaborators(Lcom/ibm/wsspi/webcontainer/collaborator/ICollaboratorMetaData;Ljava/util/EnumSet;)V
  Parent      0   0.70   1.09          48          75    J:com/ibm/wsspi/webcontainer/collaborator/CollaboratorHelper.postInvokeCollaborators(Lcom/ibm/wsspi/webcontainer/collaborator/ICollaboratorMetaData;Ljava/util/EnumSet;)V
 
    Self      0   1.08   1.72          74         118    J:com/ibm/ws/webcontainer/security/WebAppSecurityCollaboratorImpl.isCDINeeded()Z
 
   Child      0   0.10   0.32           7          22    J:java/util/Collections$UnmodifiableCollection$1.next()Ljava/lang/Object;
   Child      0   0.00   0.15           0          10    J:java/lang/Float.valueOf(Ljava/lang/String;)Ljava/lang/Float;
   Child      0   0.00   0.07           0           5    J:java/util/Collections$UnmodifiableCollection.iterator()Ljava/util/Iterator;
   Child      0   0.03   0.04           2           3    J:java/lang/String.substring(I)Ljava/lang/String;
   Child      0   0.04   0.04           3           3    J:java/util/Collections$UnmodifiableCollection$1.hasNext()Z
   Child      0   0.01   0.01           1           1    J:java/lang/String.startsWith(Ljava/lang/String;)Z
 ```

Here is the change:
https://github.com/OpenLiberty/open-liberty/commit/e4830bd16bb364cb3cd8d0bfb5a3e139d5cfe26e#diff-baa03d5f5b67c180a725b93b2a1c05f3fc0f34cfc4ced59faf3ae31f45ca8c57

Looks like we are looping through features looking for things which is taking ~1.5% longer than it did before.

### PR DESCRIPTION
**Pull request by jvanhill**
…checking whether CDI was needed for Jakarta EE9.

fixes #14600

### CODE REVIEW

File: dev/com.ibm.ws.webcontainer.security/src/com/ibm/ws/webcontainer/security/WebAppSecurityCollaboratorImpl.java [hunk 1]
```diff
@@ -461,31 +461,16 @@ public void handleException(HttpServletRequest req, HttpServletResponse rsp, Thr
...
         /*
-         * Future-proof check. Some of these combinations aren't valid, but
-         * they will not make it this far.
+         * Tried to future-proof this check by iterating over all the
+         * installed features and finding the version of appSecurity installed,
+         * but that introduced a performance degradation. So for now, checking
+         * for the appSecurity-3.0/4.0 features directly.
          */
-        return appSecurityVers >= 3.0 && specLevel >= WebContainer.SPEC_LEVEL_40;
+        return WebContainer.getServletContainerSpecLevel() >= WebContainer.SPEC_LEVEL_40 &&
```
Comment (jhanders34) [inline dev/com.ibm.ws.webcontainer.security/src/com/ibm/ws/webcontainer/security/WebAppSecurityCollaboratorImpl.java:471]: In general this whole method is an anti-pattern.  Usually this is handled with a service being applied to the instance and this not being calculated on EVERY web request.  In order to reduce the extra processing of getting the installed features I suggest to change this to the below for now in order to get us back to close to where we were.

```
        if (WebContainer.getServletContainerSpecLevel() < WebContainer.SPEC_LEVEL_40) {
            return false;
        }
        Set<String> features = provisionerService.getInstalledFeatures();
        return features.contains("appSecurity-3.0") || features.contains("appSecurity-4.0");
```

**Review by jhanders34 — CHANGES_REQUESTED** [review]

File: dev/com.ibm.ws.webcontainer.security/src/com/ibm/ws/webcontainer/security/WebAppSecurityCollaboratorImpl.java (hunk 1 above)
Comment (jvanhill) [reply dev/com.ibm.ws.webcontainer.security/src/com/ibm/ws/webcontainer/security/WebAppSecurityCollaboratorImpl.java]: Made this change.

**Comment by jvanhill** [comment]
#build

**Review by jhanders34 — APPROVED** [review]

**Comment by jvanhill** [comment]
#build
#libby

**Comment by jvanhill** [comment]
Hit another connection build failure. Rebased and rebuilding.

#build
#libby

**Comment by jvanhill** [comment]
Failed with different error:

#build

**Comment by jvanhill** [comment]
#build

**Comment by jvanhill** [comment]
#build

**Comment by jvanhill** [comment]
#libby

**Pull request merged**